server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -403,7 +406,13 @@ protected boolean processLatestReceivedCheckpoint(IndexShard replicaShard, Threa
                     latestPublishedCheckpoint
                 )
             );
-            Runnable runnable = () -> onNewCheckpoint(latestReceivedCheckpoint.get(replicaShard.shardId()), replicaShard);
+            ShardRouting primaryShard = clusterService.state().routingTable().shardRoutingTable(replicaShard.shardId()).primaryShard();
```

Comment:
Can we use the same primary node passed onto `onNewCheckpoint` method call ? I think this can be problematic with stale cluster state as we are calling `onNewCheckpoint` with primaryNode calculated from stale state.